src/shared/tools.h

```diff
@@ -140,6 +141,16 @@ static inline int bitscan(uint mask)
 #define PRINTFARGS(fmt, args)
 #endif
 
+template<class T>
+const bool inrange(std::vector<T> vec, size_t i) {
+    return i < vec.size();
+}
+
+template<class T>
+const bool inrange(std::vector<T> vec, int i) {
+    return i >= 0 && i < static_cast<int>(vec.size());
```

Comment:
You shouldn't need two specializations. Just define one generic template (maybe `static_assert`ing the use of a numeric type for the second template parameter) and cast to `decltype(vec.size())`.

In any case, narrowing `vec.size()` to `int` makes no sense, you should always cast to the larger size.

Reply:
True that, will try to solve